### ISSUE TITLE
[Detector Support]: Unable to generate Tensorrt models

### ISSUE THREAD

**Issue by Xardis** [blakeblackshear/frigate#6336]
### Describe the problem you are having

I'm attempting to generate TensoRT Models with the tensorrt-model container (build nvcr.io/nvidia/tensorrt:23.04-py3). I have followed all of the steps as listed. 

I have created a directory
I downloaded the script
The container variables are configured to point to this directory
I ran the chmod -x command

Each time the container starts, I get the error: /opt/nvidia/nvidia_entrypoint.sh: line 49: /tensorrt_models.sh: Is a directory
/opt/nvidia/nvidia_entrypoint.sh: line 49: exec: /tensorrt_models.sh: cannot execute: Is a directory

Frigate isn't loading to the point the webUI shows a camera - it just hangs. 

Frigate config file has a lot of things commented out for growth - I'm just trying to test with one camera now to get things running. Admittedly I am new to setting this up, but my gut tells me I'm following just about all of the documentation properly except maybe the config.yml file.

### Version

No error - hangs

### Frigate config file

```yaml
mqtt:
  enabled: false
  #host: <home assistant ip address>
  #user: <mqtt broker username config>
  #password: <mqtt broker password config>


go2rtc:
  streams:
    driveway_main:
      - rtsp://<hidden>:554/h264Preview_01_main
      - "ffmpeg:rtsp_cam#audio=opus"
    driveway_sub:
      - rtsp://<hidden>:554/h264Preview_01_sub
      - "ffmpeg:rtsp_cam_sub#audio=opus"
    #doorway_main:
    #  - rtsp://<username>:<password>@<ipaddress>:554/h264Preview_01_main
    #  - "ffmpeg:rtsp_cam#audio=opus"
    #doorway_sub:
    #  - rtsp://<username>:<password>@<ipaddress>:554/h264Preview_01_sub
    #  - "ffmpeg:rtsp_cam_sub#audio=opus"
cameras:
  driveway:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://<hidden>:8554/driveway_main
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://<hidden>:8554/driveway_sub
          input_args: preset-rtsp-restream
          roles:
            - detect
    record:
      enabled: false
      retain:
        days: 10
        mode: all
      events:
        pre_capture: 10
        post_capture: 20
        objects:
          - person
          - car
        retain:
          default: 15
          mode: active_objects
          objects:
            person: 15
            car: 15
    detect:
      enabled: false
      width: 1280
      height: 720
      fps: 5
    objects:
      filters:
        person:
          min_score: 0.65
          threshold: 0.75
      track:
        - person
        - car
    live:
      stream_name: rtsp_cam_sub
 # doorway:
 #   ffmpeg:
 #     output_args:
 #       record: preset-record-generic-audio-copy
 #     inputs:
 #       - path: rtsp://127.0.0.1:8554/doorway_main
 #         input_args: preset-rtsp-restream
 #         roles:
 #           - record
 #       - path: rtsp://127.0.0.1:8554/doorway_sub
 #         input_args: preset-rtsp-restream
 #         roles:
 #           - detect
 #   record:
 #     enabled: true
 #     retain:
 #       days: 10
 #       mode: all
 #     events:
 #       pre_capture: 10
 #       post_capture: 20
 #       objects:
 #         - person
 #         - car
 #       retain:
 #         default: 15
 #         mode: active_objects
 #         objects:
 #           person: 15
 #           car: 15
 #   detect:
 #     enabled: true
 #     width: 1280
 #     height: 720
 #     fps: 5
 #   objects:
 #     filters:
 #       person:
 #         min_score: 0.65
 #         threshold: 0.75
 #     track:
 #       - person
 #       - car
 #   live:
 #     stream_name: rtsp_cam_sub


detectors:
  tensorrt:
    type: tensorrt
    device: 0

birdseye:
  # Optional: Enable birdseye view (default: shown below)
  enabled: true
  # Optional: Width of the output resolution (default: shown below)
  width: 1280
  # Optional: Height of the output resolution (default: shown below)
  height: 720
  # Optional: Encoding quality of the mpeg1 feed (default: shown below)
  # 1 is the highest quality, and 31 is the lowest. Lower quality feeds utilize less CPU resources.
  quality: 8
  # Optional: Mode of the view. Available options are: objects, motion, and continuous
  #   objects - cameras are included if they have had a tracked object within the last 30 seconds
  #   motion - cameras are included if motion was detected in the last 30 seconds
  #   continuous - all cameras are included always
  mode: objects
```

### docker-compose file or Docker CLI command

```yaml
docker run
  -d
  --name='frigate'
  --net='bridge'
  -e TZ="America/Chicago"
  -e HOST_OS="Unraid"
  -e HOST_HOSTNAME="UnRaid"
  -e HOST_CONTAINERNAME="frigate"
  -e 'FRIGATE_RTSP_PASSWORD'=<hidden>
  -e 'NVIDIA_VISIBLE_DEVICES'=<hidden>
  -e 'NVIDIA_DRIVER_CAPABILITIES'='compute,utility,video'
  -l net.unraid.docker.managed=dockerman
  -l net.unraid.docker.webui='http://[IP]:[PORT:5000]'
  -l net.unraid.docker.icon='https://raw.githubusercontent.com/yayitazale/unraid-templates/main/frigate.png'
  -p '5000:5000/tcp'
  -p '8554:8554/tcp'
  -p '8555:8555/tcp'
  -p '8555:8555/udp'
  -p '1984:1984/tcp'
  -v '/mnt/user/appdata/frigate':'/config':'rw'
  -v '<hidden>':'/media/frigate':'rw'
  -v '/mnt/user/appdata/frigate/trt-models/':'/trt-models':'ro'
  -v '/etc/localtime':'/etc/localtime':'rw'
  --shm-size=256mb
  --mount type=tmpfs,target=/tmp/cache,tmpfs-size=100000000
  --restart unless-stopped
  --runtime=nvidia 'ghcr.io/blakeblackshear/frigate:stable-tensorrt'

6931cd17fcfbd39821276c912da324b9899ccd4cb472ca30880916cd352f1378
```

### Relevant log output

```shell
tenorrt-model container
/opt/nvidia/nvidia_entrypoint.sh: line 49: /tensorrt_models.sh: Is a directory
/opt/nvidia/nvidia_entrypoint.sh: line 49: exec: /tensorrt_models.sh: cannot execute: Is a directory

Frigate container

2023-04-30 22:37:44.826976732  [INFO] Preparing go2rtc config...
2023-04-30 22:37:44.828645507  [INFO] Starting Frigate...
2023-04-30 22:37:44.829126108  [INFO] Starting NGINX...
2023-04-30 22:37:44.985112485  [INFO] Starting go2rtc...
2023-04-30 22:37:45.053616869  22:37:45.053 INF go2rtc version 1.2.0 linux/amd64
2023-04-30 22:37:45.053811419  22:37:45.053 INF [api] listen addr=:1984
2023-04-30 22:37:45.054178640  22:37:45.054 INF [rtsp] listen addr=:8554
2023-04-30 22:37:45.054397641  22:37:45.054 INF [srtp] listen addr=:8443
2023-04-30 22:37:45.054535741  22:37:45.054 INF [webrtc] listen addr=:8555
2023-04-30 22:37:45.627538426  [2023-04-30 22:37:45] frigate.app                    INFO    : Starting Frigate (0.12.0-da3e197)
2023-04-30 22:37:45.648552399  [2023-04-30 22:37:45] peewee_migrate                 INFO    : Starting migrations
2023-04-30 22:37:48.131496085  2023/04/30 22:37:48 [error] 134#134: *1 connect() failed (111: Connection refused) while connecting to upstream, client: <hidden>, server: , request: "GET /api/config HTTP/1.1", upstream: "http://127.0.0.1:5001/config", host: "<hidden>", referrer: "<hidden>"
2023-04-30 22:37:48.131564246  <hidden> - - [30/Apr/2023:22:37:48 -0500] "GET /api/config HTTP/1.1" 502 157 "<hidden>" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/112.0" "-"
2023-04-30 22:37:48.598008482  <hidden> - - [30/Apr/2023:22:37:48 -0500] "GET / HTTP/1.1" 200 6641 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/112.0" "-"
2023-04-30 22:37:48.634518884  2023/04/30 22:37:48 [error] 134#134: *1 connect() failed (111: Connection refused) while connecting to upstream, client: <hidden>, server: , request: "GET /api/config HTTP/1.1", upstream: "http://127.0.0.1:5001/config", host: "<hidden>", referrer: "<hidden>"
2023-04-30 22:37:48.634570374  <hidden> - - [30/Apr/2023:22:37:48 -0500] "GET /api/config HTTP/1.1" 502 157 "<hidden>" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/112.0" "-"
2023-04-30 22:37:50.674434828  database is locked
2023-04-30 22:37:51.825081432  [INFO] Service Frigate exited with code 1 (by signal 0)
2023-04-30 22:37:51.834202795  [INFO] The go2rtc-healthcheck service exited with code 256 (by signal 15)
2023-04-30 22:37:51.872028819  [INFO] Service NGINX exited with code 0 (by signal 0)
2023-04-30 22:37:51.875347238  exit OK
2023-04-30 22:37:51.879473978  [INFO] The go2rtc service exited with code 0 (by signal 0)
2023-04-30 22:37:56.450080638  [INFO] Starting Frigate...
2023-04-30 22:37:56.450285038  [INFO] Preparing go2rtc config...
2023-04-30 22:37:56.451696242  [INFO] Starting NGINX...
2023-04-30 22:37:56.614937556  [INFO] Starting go2rtc...
2023-04-30 22:37:56.684143068  22:37:56.684 INF go2rtc version 1.2.0 linux/amd64
2023-04-30 22:37:56.684346538  22:37:56.684 INF [api] listen addr=:1984
2023-04-30 22:37:56.684729679  22:37:56.684 INF [rtsp] listen addr=:8554
2023-04-30 22:37:56.684939050  22:37:56.684 INF [srtp] listen addr=:8443
2023-04-30 22:37:56.685075780  22:37:56.685 INF [webrtc] listen addr=:8555
2023-04-30 22:37:57.295310852  [2023-04-30 22:37:57] frigate.app                    INFO    : Starting Frigate (0.12.0-da3e197)
2023-04-30 22:37:57.319223231  [2023-04-30 22:37:57] peewee_migrate                 INFO    : Starting migrations


Log window keeps closing before i can get a full copy
```

### Operating system

UNRAID

### Install method

Docker Compose

### Coral version

CPU (no coral)

### Any other information that may be helpful

GPU is Quadro P5000

**Comment by blakeblackshear** [comment]
Is there a reason you aren't following our documentation for generating models? https://docs.frigate.video/configuration/detectors#generate-models

**Comment by Xardis** [comment]
That's a funny reply. I didn't realize I wasn't. In all my time using Unraid, I haven't composed a docker container using the CLI, it's always been managed through different Community Apps - this is no different. I installed the app/container tensorrt-models which links back to this github page for support and an Unraid forum post (https://forums.unraid.net/topic/98064-support-blakeblackshear-frigate/) which I assumed to be authored by you. The username is different sure, but all roads point back here.  The instructions in this container state:

"To generate the model files with this container follow the steps:

- Create a new folder to save the models (for example on /appdata/trt-models)
- Download the script and save it in the previusly creater path (https://raw.githubusercontent.com/blakeblackshear/frigate/dev/docker/tensorrt_models.sh)
- Open the unraid console and launch the following commad without quotes(pointing to your script path): 'chmod +x /mnt/user/appdata/trt-models/tensorrt_models.sh
- Then launch the container that will run the script and stop then container at the end."

I have completed all of these steps (which mirror the steps in your documentation, minus the docker command) and then attempt to start that container and it fails with the error I pasted about. If these aren't accurate instructions that's fine - i just need to know not to follow them, or maybe I'm admittedly missing a step. I'm willing to admit I'm wrong here, but I'm asking for help where that is.

**Comment by blakeblackshear** [comment]
Someone in the community created that. I'm not an unraid user, but @NickM-27 is. Perhaps he can help figure out what you need to do.

**Comment by NickM-27** [comment]
This usually happens when your docker config is pointing to a directory that should be a file. By default docker on the first run creates any bind mounts as folders if the folder / file doesn't already exist. 

I'm not sure how the unRAID app works as I also used the cli directly when testing the TensorRT in unRAID. I'll have to look at that tomorrow if you have more questions on it.

**Comment by NickM-27** [comment]
>  This gave me an error that permission was denied. However, I've granted full access to all users, guests, etc to this file so there should not be any lack of permissions.

this sounds like you used the GUI, but the unraid GUI does not support setting execute permission in my experience. You need to do this via the command line `chmod 777 tensorrt_models.sh`

also having multiple bind mounts like

```
-v '/mnt/user/appdata/trt-models/':'/tensorrt_models':'rw'
-v '/mnt/user/appdata/trt-models/tensorrt_models.sh':'/tensorrt_models.sh':'ro'
```

is going to cause conflicts. I'd suggest deleting the second one and then walking the directory path and making sure things are as they should be.

**Comment by csreynolds** [comment]
I'd like to inject here that the docs say to use a specific container for building the models for compatibility reasons.
I ran into a problem #6092 where the frigate container expected the models to be generated with a specific version of cudalibs and did not work.

https://docs.frigate.video/configuration/detectors#generate-models :
nvcr.io/nvidia/tensorrt:22.07-py3

@Xardis docker run command: 
nvcr.io/nvidia/tensorrt:23.04-py3

**Comment by Xardis** [comment]
Thanks - I'll give it a shot. I think I started with the 22.07-py3 version and got the same directory error. I thought I'd try a newer version that might work better with the latest nvidia drivers.

**Comment by NickM-27** [comment]
To be clear the directory error is purely a config issue, the versions of anything won't affect that error

**Comment by Xardis** [comment]
I agree - I gave up on the community created container, deleted everything, and started over.

I copy/pasted the bulk command from docs and its starting to download. I have no idea why I couldn't get it to run in the folders I wanted it to. I'm gonna chalk this one up to legitimate user error mixed with some confusion from an unneeded community application. Crossing fingers my frigate container will start now and pull the camera feed.

**Comment by Xardis** [comment]
Just want to confirm - these messages pretty much mean my GPU sucks and I should buy a Coral device, right? Surprising, I haven't found much this P5000 can't do. =( 

It looked like it downloaded a lot of files, but only 4 showed up in the directory.

[05/05/2023-03:38:52] [TRT] [W] FP16 support requested on hardware without native FP16 support, performance will be negatively affected.
[05/05/2023-03:39:03] [TRT] [W] Weights [name=002_convolutional + 002_convolutional_bn.weight] had the following issues when converted to FP16:
[05/05/2023-03:39:03] [TRT] [W]  - Subnormal FP16 values detected. 
[05/05/2023-03:39:03] [TRT] [W] If this is not the desired behavior, please modify the weights or retrain with regularization to reduce the magnitude of the weights.
[05/05/2023-03:39:03] [TRT] [W] Weights [name=002_convolutional + 002_convolutional_bn.weight] had the following issues when converted to FP16:
[05/05/2023-03:39:03] [TRT] [W]  - Subnormal FP16 values detected. 
[05/05/2023-03:39:03] [TRT] [W] If this is not the desired behavior, please modify the weights or retrain with regularization to reduce the magnitude of the weights.
[05/05/2023-03:39:03] [TRT] [W] Weights [name=003_convolutional + 003_convolutional_bn || 005_convolutional + 005_convolutional_bn.weight] had the following issues when converted to FP16:
[05/05/2023-03:39:03] [TRT] [W]  - Subnormal FP16 values detected.

**Comment by NickM-27** [comment]
Those are pretty normal logs, the main way to tell if the hardware isn't up to the task is with the inference speed.

**Comment by yayitazale** [comment]
> I agree - I gave up on the community created container, deleted everything, and started over.
> 
> I copy/pasted the bulk command from docs and its starting to download. I have no idea why I couldn't get it to run in the folders I wanted it to. I'm gonna chalk this one up to legitimate user error mixed with some confusion from an unneeded community application. Crossing fingers my frigate container will start now and pull the camera feed.

This app is created by me just to help users. As you can see, the support link points to the unraid's forum of frigate, as I'm also the maintainer of the frigate template itself. 

The tensor-rt template has been tested and it is working perfectly if you just follow the instructions one by one. If you change the suggested path then there is no guarantee to work as expected. Then you launched the command manually without changing the original paths and as you see it works, so don't blame on my contribution. If you have any concern about, you are welcome to propose modifications to the template with a PR [here](https://gitlab.com/yayitazale/unraid-templates/-/blob/main/yayitazale/tensorrt-models.xml)



> Thanks Nick. I've tried configuring the script path variable a few ways.
> 
> First - i mapped to the folder that contained the script. There is a post argument configured with the scriptname that I assumed would run it - it did not. This gave me the directory error Second - I mapped the script path directly to the script itself. This gave me an error that permission was denied. However, I've granted full access to all users, guests, etc to this file so there should not be any lack of permissions.
> 
> Here is the command that is used to build the tensorrt-models.
> 
> docker run -d --name='tensorrt-models' --net='' -e TZ="America/Chicago" -e HOST_OS="Unraid" -e HOST_HOSTNAME="UnRaid" -e HOST_CONTAINERNAME="tensorrt-models" -e 'YOLO_MODELS'='yolov4-tiny-288,yolov4-tiny-416,yolov7-tiny-416' -e 'NVIDIA_VISIBLE_DEVICES'='' -e 'NVIDIA_DRIVER_CAPABILITIES'='compute,utility' -l net.unraid.docker.managed=dockerman -l net.unraid.docker.icon='https://gitlab.com/yayitazale/unraid-templates/-/raw/main/tensorrt.png' -v '/mnt/user/appdata/trt-models/':'/tensorrt_models':'rw' -v '/mnt/user/appdata/trt-models/tensorrt_models.sh':'/tensorrt_models.sh':'ro' --runtime=nvidia 'nvcr.io/nvidia/tensorrt:23.04-py3' /tensorrt_models.sh
> 
> 825cf351c4796a6092862fa1f0b5879e4175cd537c982d270e47373da202cb4c
> 
> The command finished successfully!
> 
> I also executed the script docker run --gpus=all --rm -it -v `pwd`/trt-models:/mnt/user/appdata/trt-models -v `pwd`/tensorrt_models.sh:/mnt/user/appdata/trt-models/tensorrt_models.sh nvcr.io/nvidia/tensorrt:23.04-py3 /mnt/user/appdata/trt-models/tensorrt_models.sh but received the same is a directory error message. It's like Unraid is detecting the .sh file as a folder and not a script file.

**Comment by moatus** [comment]
@yayitazale, love your Frigate work on Unraid! I am however tried following your instructions, but appear to be getting the same outcome as @Xardis.

I created the folder inside of /appdata/trt-models and then added the tensorrt_models.sh script
![image](https://github.com/blakeblackshear/frigate/assets/15126267/c9436e61-74c5-4dc8-9fa7-87e14606e0f8)

I ran `chmod +x /mnt/user/appdata/trt-models/tensorrt_models.sh` in the terminal, without any errors

I left the default locations from your template:

![image](https://github.com/blakeblackshear/frigate/assets/15126267/47526c5c-2188-4829-bfdc-26eb45e953b7)

And the result is 

![image](https://github.com/blakeblackshear/frigate/assets/15126267/9dddf018-5a64-497c-a49f-ec1fae632988)

Is there something I are missing?

**Comment by yayitazale** [comment]
> 

Can you switch to advanced view and confirm that you have this in the **Post Arguments:**?

`/tensorrt_models.sh`

**Comment by moatus** [comment]
@yayitazale, yes.

![image](https://github.com/blakeblackshear/frigate/assets/15126267/ddeeb3dd-d98b-4fb6-8e99-bbeb10699fe1)

**Comment by moatus** [comment]
@yayitazale done...

![image](https://github.com/blakeblackshear/frigate/assets/15126267/1db7b712-d864-4566-9a34-88929314e634)

but same result. Ran both `chmod` commands just to make sure.

![image](https://github.com/blakeblackshear/frigate/assets/15126267/ecd334c0-591e-4d92-9204-bcf93f13b6c5)

**Comment by yayitazale** [comment]
I see, is your filename correct? You have something wrong with the `_`

**Comment by moatus** [comment]
@yayitazale yep, I thought of that too. The filename is correct.

![image](https://github.com/blakeblackshear/frigate/assets/15126267/0a251998-dd45-49ca-a176-2b2d7ce422ec)

 Unraid's terminal just doesn't show `_`, however I did manually type it to ensure that `_` isn't being lost in translation. When you don't have `_` and instead have a space, then you get the expected error.
 
 The first has the `_` and the second has a space.
 
![image](https://github.com/blakeblackshear/frigate/assets/15126267/52f1a483-541c-49d0-9d10-0fa139899fb5)

**Comment by NickM-27** [comment]
Was the filename set via windows? They might be using different characters for the `_` because mine definitely shows the `_` correctly in unraid

**Comment by moatus** [comment]
@yayitazale Not a Windows thing. Checked in a Linux file explorer as well. Went ahead and renamed it there as well, but result is the same.

![image](https://github.com/blakeblackshear/frigate/assets/15126267/b1d92245-8e19-465b-a61e-65e34eaccd06)

Is there any issue with the script?

![image](https://github.com/blakeblackshear/frigate/assets/15126267/68ce6385-9c62-427c-85d6-8eb287b13165)

**Comment by moatus** [comment]
@NickM-27 looks like it was just a visual glitch with Firefox. On Chrome the `_` shows as it should.

![image](https://github.com/blakeblackshear/frigate/assets/15126267/8edd036a-8940-445c-be66-0d4c0bbe260b)

**Comment by francismacomber** [comment]
I recently ran into the same error where `tensorrt_models.sh` was created as a folder and no matter where I downloaded the script, the docker container wouldn't run it. These are the steps I used to get it to work:

1. Remove the second mount setting in the docker config, the one that points to the script (as mentioned here: https://github.com/blakeblackshear/frigate/issues/6336#issuecomment-1533276429)
2. Download the script to the `trt-models` folder and make it executable.
3. Switch to advanced view and change the post arguments to point to the script in the container's trt-models folder, e.g. `/tensorrt_models/tensorrt_models.sh`
4. Save and run the container, open the log and watch to confirm it runs successfully.
